layout/plan.c

```diff
@@ -0,0 +1,251 @@
+	ioo = bob_of(oo, struct m0_op_io, ioo_oo, &ioo_bobtype);
+
+	linst = oo->oo_layout_instance;
+	if (linst == NULL) {
+		M0_LOG(M0_ERROR, "layout instance is not initialised, "
+		                 "was object opened?");
+		return NULL;
+	}
+
+	M0_ALLOC_PTR(plan);
+	if (plan == NULL) {
+		M0_LOG(M0_ERROR, "failed to allocate memory for the plan");
+		return NULL;
+	}
+
+	pplops_tlist_init(&plan->lp_plops);
+	plan->lp_op = op;
+	plan->lp_layout = linst;
+
+	rc = ioo->ioo_ops->iro_iomaps_prepare(ioo) ?:
+	     ioo->ioo_nwxfer.nxr_ops->nxo_distribute(&ioo->ioo_nwxfer);
+	if (rc != 0)
+		goto out;
+
+	m0_htable_for(tioreqht, ti, &ioo->ioo_nwxfer.nxr_tioreqs_hash) {
+		plop = plop_alloc_init(plan, M0_LAT_READ, ti);
+		if (plop == NULL) {
+			rc = M0_ERR(-ENOMEM);
+			break;
+		}
+		plop->pl_ent = ti->ti_fid;
+		iopl = container_of(plop, struct m0_layout_io_plop, iop_base);
+		iopl->iop_ext  = ti->ti_ivec;
+		iopl->iop_data = ti->ti_bufvec;
+		iopl->iop_session = ti->ti_session;
+		iopl->iop_goff = ti->ti_goff;
+	} m0_htable_endfor;
+
+ out:
+	if (rc != 0) {
+		m0_layout_plan_fini(plan);
+		plan = NULL;
+	}
+
+	M0_LEAVE();
+	return plan;
+}
+
+M0_INTERNAL void m0_layout_plan_fini(struct m0_layout_plan *plan)
+{
+	struct m0_layout_plop  *plop;
+	struct m0_op_common    *oc;
+	struct m0_op_obj       *oo;
+	struct m0_op_io        *ioo;
+	struct target_ioreq    *ti;
+
+	M0_ENTRY("plan=%p", plan);
+
+	oc = bob_of(plan->lp_op, struct m0_op_common, oc_op, &oc_bobtype);
+	oo = bob_of(oc, struct m0_op_obj, oo_oc, &oo_bobtype);
+	ioo = bob_of(oo, struct m0_op_io, ioo_oo, &ioo_bobtype);
+
+	m0_htable_for(tioreqht, ti, &ioo->ioo_nwxfer.nxr_tioreqs_hash) {
+		tioreqht_htable_del(&ioo->ioo_nwxfer.nxr_tioreqs_hash, ti);
+		target_ioreq_fini(ti);
+	} m0_htable_endfor;
+
+	if (ioo->ioo_iomaps != NULL)
+		ioo->ioo_ops->iro_iomaps_destroy(ioo);
+
+	m0_tl_teardown(pplops, &plan->lp_plops, plop) {
+		if (plop->pl_ops && plop->pl_ops->po_fini)
+			plop->pl_ops->po_fini(plop);
+		/* for each plan_get() plop_done() must be called */
+		M0_ASSERT(M0_IN(plop->pl_state, (M0_LPS_INIT, M0_LPS_DONE)));
+		m0_free(plop);
+	}
+	pplops_tlist_fini(&plan->lp_plops);
+	plan->lp_op = NULL;
+	m0_free(plan);
+
+	M0_LEAVE();
+}
+
+M0_INTERNAL int m0_layout_plan_get(struct m0_layout_plan *plan, uint64_t colour,
+				   struct m0_layout_plop **plop)
+{
+	if (plan == NULL || plop == NULL)
+		return M0_ERR(-EINVAL);
+
+	if (plan->lp_op == NULL)
+		return M0_ERR_INFO(-EINVAL, "plan %p is not built", plan);
+
+	if (plan->lp_last_plop == NULL)
+		*plop = pplops_tlist_tail(&plan->lp_plops);
+	else
+		*plop = pplops_tlist_prev(&plan->lp_plops, plan->lp_last_plop);
+
+	if (*plop == NULL) {
+		*plop = plop_alloc_init(plan, M0_LAT_DONE, NULL);
+		if (*plop == NULL)
+			return M0_ERR(-ENOMEM);
+	}
+	plan->lp_last_plop = *plop;
+
+	return M0_RC(0);
+}
+
+M0_INTERNAL int m0_layout_plop_start(struct m0_layout_plop *plop)
+{
+	if (plop->pl_state != M0_LPS_INIT)
+		return M0_ERR(-EINVAL);
+
```

Comment:
Also

M0_ASSERT(m0_tl_forall(pplops, scan, plop->pl_deps, scan->pl_state == M0_LPS_DONE));

Reply:
Added, thanks.

Reply:
m0_layout_plop_start() should be mandatory at least to keep the interface simple.